cmd/fpga_admissionwebhook/patcher_test.go

```diff
@@ -0,0 +1,353 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"testing"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	fpgav1 "github.com/intel/intel-device-plugins-for-kubernetes/pkg/apis/fpga.intel.com/v1"
+)
+
+func TestPatcherStorageFunctions(t *testing.T) {
+	af := &fpgav1.AcceleratorFunction{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "arria10-nlb0",
+		},
+		Spec: fpgav1.AcceleratorFunctionSpec{
+			InterfaceID: "ce48969398f05f33946d560708be108a",
+			AfuID:       "d8424dc4a4a3c413f89e433683f9040b",
+		},
+	}
+	region := &fpgav1.FpgaRegion{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "arria10",
+		},
+		Spec: fpgav1.FpgaRegionSpec{
+			InterfaceID: "ce48969398f05f33946d560708be108a",
+		},
+	}
+
+	p, err := newPatcher(preprogrammed)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	p.addAf(af)
+	if len(p.afMap) != 1 || len(p.resourceMap) != 1 {
+		t.Error("Failed to add AF to patcher")
+	}
+
+	p.removeAf(af.Name)
+	if len(p.afMap) != 0 || len(p.resourceMap) != 0 {
+		t.Error("Failed to remove AF from patcher")
+	}
+
+	p.addRegion(region)
+	if len(p.regionMap) != 1 || len(p.resourceMap) != 1 {
+		t.Error("Failed to add fpga region to patcher")
+	}
+
+	p.removeRegion(region.Name)
+	if len(p.regionMap) != 0 || len(p.resourceMap) != 0 {
+		t.Error("Failed to remove fpga region from patcher")
+	}
+}
+
+func TestParseResourceName(t *testing.T) {
+	tcases := []struct {
+		input       string
+		interfaceID string
+		afuID       string
+		afMap       map[string]string
+		regionMap   map[string]string
+		expectedErr bool
+	}{
+		{
+			input: "fpga.intel.com/arria10",
+			regionMap: map[string]string{
+				"arria10": "ce48969398f05f33946d560708be108a",
+			},
+			interfaceID: "ce48969398f05f33946d560708be108a",
+		},
+		{
+			input: "fpga.intel.com/arria10-unknown",
+			regionMap: map[string]string{
+				"arria10": "ce48969398f05f33946d560708be108a",
+			},
+			expectedErr: true,
+		},
+		{
+			input:       "fpga.intel.com/unknown",
+			expectedErr: true,
+		},
+		{
+			input: "fpga.example.com/something",
+		},
+		{
+			input: "fpga.intel.com/arria10-nlb0",
+			regionMap: map[string]string{
+				"arria10": "ce48969398f05f33946d560708be108a",
+			},
+			afMap: map[string]string{
+				"arria10-nlb0": "d8424dc4a4a3c413f89e433683f9040b",
+			},
+			interfaceID: "ce48969398f05f33946d560708be108a",
+			afuID:       "d8424dc4a4a3c413f89e433683f9040b",
+		},
```

Comment:
case for unknown afu id ?

Reply:
It's in the line 89

Reply:
I meant something like "unknown-nlb0" variant.